File: runtime-elixir/server/smithy_server.ex
defmodule SmithyServer do
  @moduledoc """
  Plug-compatible HTTP abstraction layer for Smithy-generated Elixir servers.

  Generated dispatcher modules call these helpers to extract request data and
  build responses, keeping protocol details out of service business logic.

  Mirrors `smithy_server.erl` from the Erlang runtime.

  ## Usage in a generated dispatcher

      def call(%Plug.Conn{} = conn, _opts) do
        {method, path, headers, body} = SmithyServer.extract(conn)
        case route(method, path, headers, body) do
          {:ok, result}    -> SmithyServer.response(conn, 200, result)
          {:error, err}    -> SmithyServer.error_response(conn, err)
          :not_found       -> SmithyServer.not_found(conn)
        end
      end
  """

  @doc """
  Extract method, path, headers, and body from a `Plug.Conn`.

  Returns `{method, path, headers, body}` where:

    * `method` — uppercase HTTP method string, e.g. `"POST"`.
    * `path` — request path string, e.g. `"/items/42"`.
    * `headers` — request headers as a `%{String.t() => String.t()}` map
      (all names lowercased, matching Plug conventions).
    * `body` — raw request body binary.
  """
  @spec extract(Plug.Conn.t()) ::
          {method :: String.t(), path :: String.t(), headers :: map(), body :: binary()}
  def extract(%Plug.Conn{} = conn) do
    {:ok, body, _conn} = Plug.Conn.read_body(conn)
    headers = conn.req_headers |> Map.new()
    {conn.method, conn.request_path, headers, body}
  end

  @doc """
  Send a successful JSON response.

  Sets `content-type: application/json` and serialises `body` with `Jason`.
  """
  @spec response(Plug.Conn.t(), pos_integer(), term()) :: Plug.Conn.t()
  def response(conn, code, body) do
    conn
    |> Plug.Conn.put_resp_content_type("application/json")
    |> Plug.Conn.send_resp(code, Jason.encode!(body))
  end

  @doc """
  Send a JSON error response for a modelled Smithy error term.

  Delegates HTTP status code and message lookup to `SmithyErrorMap.to_http/1`.
  """
  @spec error_response(Plug.Conn.t(), term()) :: Plug.Conn.t()
  def error_response(conn, err) do
    {code, msg} = SmithyErrorMap.to_http(err)
    response(conn, code, %{message: msg})
  end

  @doc """
  Send a 400 Bad Request response for a validation failure.

  `reason` must be a term accepted by `SmithyValidator.format/1`.
  """
  @spec validation_error(Plug.Conn.t(), term()) :: Plug.Conn.t()
  def validation_error(conn, reason) do
    response(conn, 400, %{message: SmithyValidator.format(reason)})
  end

  @doc """
  Send a 404 Not Found response.
  """
  @spec not_found(Plug.Conn.t()) :: Plug.Conn.t()
  def not_found(conn) do
    Plug.Conn.send_resp(conn, 404, "Not Found")
  end
end

File: runtime-erlang/server/smithy_server.erl
-module(smithy_server).

%% HTTP abstraction layer for Smithy server-side dispatch.
%% Framework-agnostic API; the Cowboy implementation is provided here.
%% The return type for response helpers is {StatusCode, Headers, Body} —
%% a plain tuple. The application's Cowboy handler converts it to a
%% framework-specific response.

-export([
    extract/1,
    response/2,
    error_response/1,
    validation_error/1,
    not_found/0
]).

%%%===================================================================
%%% API Functions
%%%===================================================================

%% @doc Extract method, path, headers, and body from a Cowboy request.
%% Returns {Method, Path, Headers, Body} where all values are binaries.
-spec extract(Req :: cowboy_req:req()) ->
    {Method :: binary(), Path :: binary(), Headers :: cowboy:http_headers(), Body :: binary()}.
extract(Req) ->
    Method  = cowboy_req:method(Req),
    Path    = cowboy_req:path(Req),
    Headers = cowboy_req:headers(Req),
    {ok, Body, _Req2} = cowboy_req:read_body(Req),
    {Method, Path, Headers, Body}.

%% @doc Build a successful response tuple with JSON content-type.
%% Returns {StatusCode, Headers, Body}.
-spec response(Code :: pos_integer(), Body :: iodata()) ->
    {pos_integer(), [{binary(), binary()}], iodata()}.
response(Code, Body) ->
    {Code, [{<<"content-type">>, <<"application/json">>}], Body}.

%% @doc Build an error response from a modeled error term.
%% Delegates HTTP status and message lookup to smithy_error_map.
-spec error_response(Err :: term()) ->
    {pos_integer(), [{binary(), binary()}], binary()}.
error_response(Err) ->
    {Code, Msg} = smithy_error_map:to_http(Err),
    Body = jsx:encode(#{<<"message">> => Msg}),
    {Code, [{<<"content-type">>, <<"application/json">>}], Body}.

%% @doc Build a 400 validation-error response.
%% Reason is a term accepted by smithy_validator:format/1.
-spec validation_error(Reason :: term()) ->
    {400, [{binary(), binary()}], binary()}.
validation_error(Reason) ->
    Body = jsx:encode(#{<<"message">> => smithy_validator:format(Reason)}),
    {400, [{<<"content-type">>, <<"application/json">>}], Body}.

%% @doc Build a 404 Not Found response.
-spec not_found() -> {404, [], binary()}.
not_found() ->
    {404, [], <<"Not Found">>}.
